docs/adrs/0080-use-flipt-feature-flags.md

```diff
@@ -0,0 +1,164 @@
+
+Rolling our own is a possibility, but one key feature we'd want to think about is how to promote feature flags from one environment to another (e.g. from staging to production). That suggests that tying the feature flag to the environment by e.g. storing the flags in the database isn't ideal as we need to recreate the flag settings in each environment. Combining that with the desire to have feature flags be enabled on a per user basis, it makes the complexity of rolling our own outweigh the relatively simple flipt deployment.
+
+Flipt provides a [filesystem backend](https://www.flipt.io/docs/experimental/filesystem-backends) which would allow a way for us to manage our feature flags using a [gitops](https://about.gitlab.com/topics/gitops/) style process. We can test out our flag configuration in separate environments (e.g. experimental, demo, staging) before rolling out to production. It also allows us to deploy the Flipt service without requiring another stateful system (e.g. no database).
+
+We will have a couple of different options for how we deploy flipt. Examining the options for how flipt is deployed will be its own ADR.
+
+## Pros and Cons of the Alternatives
+
+### Option #0: Do Nothing
+* '+' No change to App Eng workflow
+* '+' No additional work for Platform Team
+* '-' We will at some point ship the wrong code that includes a half baked feature that wasn't meant for public use yet
+* '-' Could break API versioning required by Prime and/or make it harder to make API changes
+
+
+### Option #1: [LaunchDarkly](https://launchdarkly.com)
+* '+' Industry standard
+* '+' Feature Rich
+* '-' Expensive. No room in ODC budget
+* '-' Might not be allowed (It is FedRAMP moderate, not IL4)
+* '-' Unknown if Platform Team will need to be involved
+
+
+### Option #2: [Unleash](https://github.com/Unleash/unleash)
+* '+' Open Source
+* '+' Can be self-hosted, so meets MilMove security requirements
+* '-' Adds another thing for Platform to do before Milestone 1
+* '-' Open source version may be limited
+* '-' Somewhat complicated deployment model
+
+
+### Option #3: [AWS AppConfig](https://docs.aws.amazon.com/appconfig/latest/userguide/what-is-appconfig.html)
+* '+' Not a lot of additional setup, App Eng already used to getting other configuration information from AWS
+* '+' Very little Platform setup, AppEng already has limited access to AWS console
+* '+' No separate approval, meets MilMove security requirements
+* '-' Unknown cost
+* '-' Very low level
+
+### Option #4: Roll our own
+* '+' Provides all flexibility we want
+* '+' Easy to deploy (probably?) and meets MilMove security requirements
+* '-' No easy way to migrate settings from one environment to another
+* '-' We have to implement all the flexibility
+
+### Option #5: [Flipt](https://www.flipt.io)
+* '+' Provides flexibility to enable feature flags to groups of users in almost any configuration we could imagine
```

Comment:
Do we know yet how we might cross reference a user to their segment or flipt configuration from say a logging perspective?  Will we add some new fields to the context to understand what was active at a certain time?

Reply:
This is a good question. We will have to see what we need in practice. I could see logging the response for every feature flag query (which would also have the `session_id`) which would seem to provide a way to work out what happened.

We might also find on the server side we want to do something like

```
if flag == "thing_one" {
  logger.Info("Using thing_one flag")
} else {
  logger.Info("Not using thing_one flag")
}

Client side it's a bit tricker, huh?